draft-ietf-webpush-encryption.md

```diff
@@ -108,170 +101,277 @@ encryption scheme by applications that use push messaging.  An agent that only
+   context = "P-256" || 0x00 ||
+             0x00 || 0x41 || ua_public ||
+             0x00 || 0x41 || as_public
+   salt = random(16)
+   PRK = HMAC-SHA-256(salt, IKM)
+   cek_info = "Content-Encoding: aesgcm" || 0x00 || context
+   CEK = HMAC-SHA-256(PRK, cek_info || 0x01)[0..15]
+   nonce_info = "Content-Encoding: nonce" || 0x00 || context
+   NONCE = HMAC-SHA-256(PRK, nonce_info || 0x01)[0..11]
+~~~
 
-The Application Server then encrypts the payload.  Header fields are populated
-with base64url encoded [RFC7515] values:
+Note that this omits the exclusive OR of the final nonce with the record
+sequence number, since implementations of this specification are not required to
+support multiple records (see {{restrict}}) and the sequence number of the first
+record is zero.
 
-* the salt is added to the `salt` parameter of the Encryption header field; and
 
-* the public key for its DH or ECDH key pair is placed in the `dh` parameter of
-  the Crypto-Key header field.
+# Restrictions on Use of "aesgcm" Content Coding {#restrict}
 
-An application server MUST encrypt a push message with a single record.  This
+An Application Server MUST encrypt a push message with a single record.  This
 allows for a minimal receiver implementation that handles a single record.  If
 the message is 4096 octets or longer, the `rs` parameter MUST be set to a value
 that is longer than the encrypted push message length.
 
-Note that a push service is not required to support more than 4096 octets of
-payload body, which equates to 4077 octets of cleartext, so the `rs` parameter
-can be omitted for messages that fit within this limit.
+ A push service is not required to support more than 4096 octets of payload body
+(see Section 7.2 of {{!I-D.ietf-webpush-protocol}}), which equates to 4077
+octets of cleartext, so the `rs` parameter can be omitted for messages that fit
+within this limit.
 
-An application server MUST NOT use other content encodings for push messages.
+An Application Server MUST NOT use other content encodings for push messages.
 In particular, content encodings that compress could result in leaking of push
 message contents.  The Content-Encoding header field therefore has exactly one
-value, which is `aesgcm128`.  Multiple `aesgcm128` values are not permitted.
+value, which is `aesgcm`.  Multiple `aesgcm` values are not permitted.
 
-An application server MUST include exactly one entry in the Encryption field,
+An Application Server MUST include exactly one entry in the Encryption field,
 and at most one entry having a `dh` parameter in the Crypto-Key field. This
 allows the `keyid` parameter to be omitted from both header fields.
 
-An application server MUST NOT include an `aesgcm128` parameter in the
-Encryption header field.
+An Application Server MUST NOT include an `aesgcm` parameter in the Encryption
+header field.
 
+A User Agent is not required to support multiple records.  Such a User Agent
```

Comment:
This is much lighter than the MUST in restrictions as well. If an AS MUST use a single record, why would we merely suggest that UAs don't have to support multiple records? For purposes of Web Push, they MUST only accept a single record.